src/dmd/optimize.d

```diff
@@ -540,67 +540,69 @@ Expression Expression_optimize(Expression e, int result, bool keepLvalue)
         }
         if (auto ae = e.e1.isIndexExp())
         {
-            // Convert &array[n] to &array+n
-            if (ae.e2.isIntegerExp() && ae.e1.isVarExp())
-            {
-                sinteger_t index = ae.e2.toInteger();
-                VarExp ve = ae.e1.isVarExp();
-                if (ve.type.isTypeSArray() && !ve.var.isImportedSymbol())
-                {
-                    TypeSArray ts = ve.type.isTypeSArray();
-                    sinteger_t dim = ts.dim.toInteger();
-                    if (index < 0 || index >= dim)
-                    {
-                        /* 0 for C static arrays means size is unknown, no need to check
-                         */
-                        if (!(dim == 0 && ve.var.isCsymbol()))
-                        {
-                            e.error("array index %lld is out of bounds `[0..%lld]`", index, dim);
-                            return error();
-                        }
-                    }
+			if (auto ie = ae.e2.isIntegerExp())
+			{
+				// Convert &array[n] to &array+n
+				if (auto ve = ae.e1.isVarExp())
+				{
+					if (ve && ve.type.isTypeSArray() && !ve.var.isImportedSymbol())
+					{
+						TypeSArray ts = ve.type.isTypeSArray();
```

Comment:
This call is also duplicated, however, if you were to eliminate it, then the indentation would increase further. Therefore, I am at doubt if this refactoring is actually improving anything.